### DISCUSSION TITLE
Relationship problems

### DISCUSSION THREAD

**Discussion by julianadormon** [Azure/data-api-builder#1466] [General]
I am converting to DAB from Entity Framework.
I am trying to replicate the following query using graphql:

var GalleryTheme = await (from e in db.Events
                                                  from t in db.GalleryThemes
                                                  where e.EventId == PageVm.EventId
                                                  && t.GalleryThemeId == e.GalleryThemeId
                                                  select new
                                                  {
                                                      OverlayFont = t.OverlayFont,
                                                      OverlayStripFont = t.OverlayStripFont,
                                                      OverlayMenuFont = t.OverlayMenuFont,
                                                      OverlayButtonFont = t.OverlayButtonFont,
                                                      ThumbnailMenuFont = t.ThumbnailMenuFont
                                                  }).AsNoTracking().SingleOrDefaultAsync();
                                                  
with:

query GetGalleryThemeByEventId($input: Int!) {
  event_by_pk(EventId: $input) {
    EventId
    GalleryTheme {
        OverlayFont
        OverlayStripFont
        OverlayMenuFont
        OverlayButtonFont
        ThumbnailMenuFont
    }
  }
}

I created the relationships as follows:
Events can be linked to one GalleryTheme and a GalleyTheme can be linked to many Events.

dab update Event --relationship "GalleryTheme" --target.entity "GalleryTheme" --cardinality one --relationship.fields "GalleryThemeId:GalleryThemeId"

dab update GalleryTheme --relationship "Event" --target.entity "Event" --cardinality many --relationship.fields "GalleryThemeId:GalleryThemeId"

When building the project, the following error occurs:
The field `GalleryTheme` does not exist on the type `Event`.

I believe this is a Strawbery Shake error which I am using as a graphql client.

Also 
1. How to remove a relationship or entity using DAB CLI?
2. How is schema.graphql generated?
3. Are their ways to force an update to it?

Thanks?

**Comment by seantleonard** [comment]
> When building the project, the following error occurs

You mention the project build fails; do you mean that starting the Data Api builder engine fails, or that a CLI error is emitted?
I'm curious whether you are able to start the engine and get the BananaCakePop UI in order to analyze the generated GraphQL schema in order to help troubleshoot further.

1. Currently, you can only remove the relationship/entity from within the JSON file itself. We are tracking the 'remove' functionality in the CLI here: https://github.com/Azure/data-api-builder/issues/642
2. The GraphQL schema is generated during the DAB engine startup. 
    - Schema creation is based on both database metadata determined from the target DB in the connection string, and then metadata in the DAB runtime config.
4. Restarting the engine with new DAB runtime config would regenerate the GraphQL engine.

**Comment by julianadormon** [comment]
Thanks! I had to restart DAB to get the errors to go away. Is that normal?
Seems like I have to resart Visual Studio quite a bit too.

**Reply by seantleonard** [reply]
It would be normal to restart the engine to get any configuration updates to be recognized. We are aware of the ask to add "hot reload" of config: https://github.com/Azure/data-api-builder/issues/67

However, it is not normal that you'd need to restart Visual Studio. What errors are you seeing with Visual Studio?

**Reply by julianadormon** [reply]
Ok good to know, thanks. The VS reboot seems necessary sometimes to see the Strawberry shake classes, so that is not on you. Thanks again for all your speedy replies. I seem to be getting a handle on this.

**Comment by Aniruddh25** [comment]
hi @julianadormon, 

After adding the relationships, you can use nested filtering to filter on the GalleryTheme based on the event id 1234 (for eg) as follows:

> {
>   GalleryTheme (
>     filter: { 
>       Event : {
>         EventId: { eq: 1234 }
>       }
>     })  {
>     items {
>       OverlayFont
>     OverlayStripFont
>     OverlayMenuFont
>     OverlayButtonFont
>     ThumbnailMenuFont
>     }
>   }
> }
